pkg/operator/etcdservicecontroller/etcdservice_controller.go

```diff
@@ -0,0 +1,186 @@
+package etcdservicecontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+var (
+	controllerWorkQueueKey = "key"
+
+	degradedConditionNames = []string{
+		"EtcdServiceDegraded",
+		"EtcdEndpointDegraded",
+	}
+)
+
+type EtcdServiceController struct {
+	targetNamespace string
+
+	operatorClient  v1helpers.OperatorClient
+	endpointsGetter corev1client.EndpointsGetter
+	servicesGetter  corev1client.ServicesGetter
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+}
+
+func NewEtcdServiceController(
+	targetNamespace string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	servicesGetter corev1client.ServicesGetter,
+	endpointsGetter corev1client.EndpointsGetter,
+	eventRecorder events.Recorder,
+) *EtcdServiceController {
+	c := &EtcdServiceController{
+		targetNamespace: targetNamespace,
+		operatorClient:  operatorClient,
+		servicesGetter:  servicesGetter,
+		endpointsGetter: endpointsGetter,
+		eventRecorder:   eventRecorder.WithComponentSuffix("etcd-service-controller"),
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EtcdServiceController"),
+	}
+
+	operatorClient.Informer().AddEventHandler(c.eventHandler())
+
+	kubeInformersForTargetNamespace.Core().V1().Services().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
+
+	c.cachesToSync = append(c.cachesToSync, operatorClient.Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Services().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Endpoints().Informer().HasSynced)
+
+	return c
+}
+
+func (c *EtcdServiceController) sync() error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	if _, err := c.servicesGetter.Services(c.targetNamespace).Get("etcd", metav1.GetOptions{}); err != nil {
+		foundConditions = append(foundConditions, operatorv1.OperatorCondition{
```

Comment:
I think I like the immediate "set it now" behavior in https://github.com/openshift/cluster-openshift-apiserver-operator/pull/301 .  What do you think in comparison?  it makes the pairs very obvious.